File: ruby/lua_deflate.rb
class LuaDeflate
  # Define a custom base64 character set.
  BYTE_TO_BIT = [
    *'a'..'z',
    *'A'..'Z',
    *'0'..'9',
    '(',
    ')'
  ].freeze

  # Create a reverse lookup hash from the custom base64 character set.
  BIT_TO_BYTE = BYTE_TO_BIT.each.with_index.to_a.to_h.freeze

  class << self
    def decode_for_print(encoded_str)
      return unless encoded_str.is_a?(String)

      # Strip leading and trailing whitespace.
      encoded_str = encoded_str.strip
      return if encoded_str.length <= 1

      decoded_bytes = encoded_str.chars.each_slice(4).flat_map do |char_group|
        # Convert each character in the chunk to its corresponding base64 index.
        indices = char_group.map { |char| BIT_TO_BYTE[char] }
        return nil if indices.include?(nil)

        # Calculate the 24-bit number represented by the chunk.
        value = indices.each_with_index.sum { |index, idx| index * (64**idx) }

        # Determine the number of bytes this chunk should produce.
        bytes_to_take = char_group.length == 4 ? 3 : char_group.length - 1

        # Extract bytes from the 24-bit number.
        bytes_to_take.times.map { |shift| ((value >> (8 * shift)) & 0xFF).chr }
      end

      decoded_bytes.join
    end

    def encode_for_print(str)
      raise ArgumentError, "Expected 'str' to be a string, got #{str.class}" unless str.is_a?(String)

      encoded_chunks = str.chars.each_slice(3).map do |byte_group|
        # Convert characters to their ASCII byte values.
        bytes = byte_group.map(&:ord)

        # Calculate the 24-bit number represented by the 3 bytes.
        value = bytes.each_with_index.sum { |byte, idx| byte * (256**idx) }

        # Determine the number of chunks this group should produce.
        chunks_to_take = byte_group.length + 1

        # Extract 6-bit chunks from the 24-bit number and convert them to characters.
        chunks_to_take.times.map { |idx| BYTE_TO_BIT[(value >> (6 * idx)) & 0x3F] }
      end

      encoded_chunks.flatten.join
    end
  end
end

File: js/lib/LuaDeflate.js
class LuaDeflate {
    constructor() {
        // Define a custom base64 character set.
        this.BYTE_TO_BIT = [...'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789()'];

        // Create a reverse lookup map from the custom base64 character set.
        this.BIT_TO_BYTE = new Map(this.BYTE_TO_BIT.map((val, idx) => [val, idx]));
    }

    decodeForPrint(encodedStr) {
        if (typeof encodedStr !== "string") {
            return;
        }

        // Strip leading and trailing whitespace.
        encodedStr = encodedStr.trim();
        if (encodedStr.length <= 1) {
            return;
        }

        const decodedBytes = [];
        for (let i = 0; i < encodedStr.length; i += 4) {
            const charGroup = encodedStr.slice(i, i + 4).split('');

            // Convert each character in the chunk to its corresponding base64 index.
            const indices = charGroup.map(char => this.BIT_TO_BYTE.get(char));
            if (indices.includes(undefined)) {
                return null;
            }

            // Calculate the 24-bit number represented by the chunk.
            const value = indices.reduce((acc, index, idx) => acc + index * (64 ** idx), 0);

            // Determine the number of bytes this chunk should produce.
            const bytesToTake = (charGroup.length === 4) ? 3 : charGroup.length - 1;

            // Extract bytes from the 24-bit number.
            for (let shift = 0; shift < bytesToTake; shift++) {
                const byte = String.fromCharCode((value >> (8 * shift)) & 0xFF);
                decodedBytes.push(byte);
            }
        }

        return decodedBytes.join('');
    }

    decodeForPrint2(encodedStr) {
        if (typeof encodedStr !== "string") {
            return;
        }

        encodedStr = encodedStr.trim();
        if (encodedStr.length <= 1) {
            return;
        }

        const byteCount = Math.ceil(encodedStr.length * 6 / 8);
        const buffer = new ArrayBuffer(byteCount);
        const view = new Uint8Array(buffer);

        let bufferIndex = 0;
        for (let i = 0; i < encodedStr.length; i += 4) {
            const charGroup = encodedStr.slice(i, i + 4).split('');

            const indices = charGroup.map(char => this.BIT_TO_BYTE.get(char));
            if (indices.includes(undefined)) {
                return null;
            }

            const value = indices.reduce((acc, index, idx) => acc + index * (64 ** idx), 0);

            const bytesToTake = (charGroup.length === 4) ? 3 : charGroup.length - 1;
            for (let shift = 0; shift < bytesToTake; shift++) {
                view[bufferIndex++] = (value >> (8 * shift)) & 0xFF;
            }
        }

        return new Uint8Array(buffer)
    }

    encodeForPrint(str) {
        if (typeof str !== "string") {
            throw new TypeError(`Expected 'str' to be a string, got ${typeof str}`);
        }

        const encodedChunks = [];
        for (let i = 0; i < str.length; i += 3) {
            const byteGroup = str.slice(i, i + 3).split('');

            // Convert characters to their ASCII byte values.
            const bytes = byteGroup.map(char => char.charCodeAt(0));

            // Calculate the 24-bit number represented by the 3 bytes.
            const value = bytes.reduce((acc, byte, idx) => acc + byte * (256 ** idx), 0);

            // Determine the number of chunks this group should produce.
            const chunksToTake = byteGroup.length + 1;

            // Extract 6-bit chunks from the 24-bit number and convert them to characters.
            for (let idx = 0; idx < chunksToTake; idx++) {
                const chunk = this.BYTE_TO_BIT[(value >> (6 * idx)) & 0x3F];
                encodedChunks.push(chunk);
            }
        }

        return encodedChunks.join('');
    }
}

// export default LuaDeflate;
module.exports = LuaDeflate;